### PR TITLE
Product variants

### PR DESCRIPTION
**Pull request by bricht**
### Related Pull Requests
https://github.com/magento/commerce-data-export/pull/8
https://github.com/magento-commerce/graphql/pull/4
https://github.com/magento-commerce/magento2-infrastructure/pull/1368
https://git.corp.adobe.com/ecp/api-protos/pull/22
https://github.com/duhon/magento-docker/pull/30

### Fixed Issues (if relevant)

1. Fixes magento/catalog-storefront#331
2. Fixes magento/catalog-storefront#332
3. Fixes magento/catalog-storefront#335

### Code Review Checklist (*)

See detailed [checklist](https://github.com/magento/catalog-storefront/blob/develop/dev/docs/projectAgreements/Code-Review-checklist.md)

- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
  - expected results in test verified with data from fixture
- [ ] no backward incompatible changes
- [ ] Export API (et_schema.xml) and SF API schemas (proto schema) are reflected in the codebase
  - prerequisite: story branch created with all needed generated classes according to proposes schema-changes
  - DTO classes do not contain any manual changes (Magento\CatalogExportApi\*, Magento\CatalogStorefrontApi\*)
- [ ] Class usage: magento/catalog-storefront repo don't use directly classes from magento/saas-export repo and vise-verse
  - Check composer.json dependencies
- [ ] Legacy code is deleted
  - Any Data Providers present in Connector part  (Magento\CatalogStorefrontConnector, Magento\*Extractor modules)
  - And Data Providers from Export API (magento/saas-export repo) that is not relevant anymore
  - Any DTO for Export API/SF API which does not reflect current schema: et_schema, proto schema
  - Any “mapper” on Message Broker (between Export API and SF API)
    - if mapper still needed, verify fields used in mapping, remove not relevant fields

### CODE REVIEW

**Comment by bricht** [comment]
@magento run all tests
without extensions magento/saas-export, magento/data-solutions-services-id, magento/magento-services-connector

**Comment by bricht** [comment]
@magento run all tests
without extensions magento/saas-export, magento/data-solutions-services-id, magento/magento-services-connector

**Comment by bricht** [comment]
@magento run all tests
without extensions magento/saas-export, magento/data-solutions-services-id, magento/magento-services-connector

**Comment by bricht** [comment]
@magento run Integration Tests
without extensions magento/saas-export, magento/data-solutions-services-id, magento/magento-services-connector

File: app/code/Magento/CatalogStorefront/Model/VariantService.php [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+     *
+     * Only variants whose corresponding products are 'enabled' and stored in storage are returned.
+     *
+     * @param ProductVariantRequestInterface $request
+     * @return ProductVariantResponseInterface
+     * @throws \InvalidArgumentException
+     * @throws RuntimeException
+     * @throws \Throwable
+     */
+    public function GetProductVariants(ProductVariantRequestInterface $request): ProductVariantResponseInterface
```
Comment (RuslanKostiv1) [inline app/code/Magento/CatalogStorefront/Model/VariantService.php:213]: I have created configurable product with 2 options:
Color: red, green, blue
Size: s, m, l

result of this method is incorrect

image

Each variation should have 2 option values, but as we see on the screenshot only half of them have 2 options, the rest of them have 1 option value for some reason.

For id `configurable/32/36` there's only one option value `32:color/Y29uZmlndXJhYmxlLzkzLzE0` however in storage we have 2 records 

image

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+     * @throws \Throwable
+     */
+    public function testSaveAndDeleteProductVariant(): void
+    {
+        $configurable = $this->getProduct(self::CONFIGURABLE_SKU);
+        $configurableId = $configurable->getId();
+        $simple1 = $this->getProduct(self::SIMPLE1_SKU);
+        $simple1Id = $simple1->getId();
+        $simple2 = $this->getProduct(self::SIMPLE2_SKU);
+        $simple2Id = $simple2->getId();
```
Comment (RuslanKostiv1) [inline dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php:107]: add more variants to configurable product

**Review by RuslanKostiv1 — CHANGES_REQUESTED** [review]

**Comment by RuslanKostiv1** [comment]
@magento run all tests

File: app/code/Magento/CatalogStorefront/Model/Storage/State.php [hunk 3]
```diff
@@ -41,6 +41,8 @@ public function getAliasName(): string
     /**
      * Get current data source name of storage taking into account version of the data source.
      *
+     * TODO: Adapt to work without store code
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/Storage/State.php]: Done

File: app/code/Magento/CatalogStorefront/Test/Api/ProductVariants/ConfigurableVariantsTest.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+     * @throws NoSuchEntityException
+     * @throws \Throwable
+     */
+    public function testConfigurableProductVariants(): void
+    {
+        /** @var $configurable Product */
+        $configurable = $this->productRepository->get('configurable');
+        $simples = [
+            $this->productRepository->get('simple_10'),
+            $this->productRepository->get('simple_20')
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Test/Api/ProductVariants/ConfigurableVariantsTest.php]: Done

File: app/code/Magento/CatalogStorefront/Model/VariantService.php (hunk 1 above)
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/VariantService.php:214]: Was a bug with max search size. Changed it to 2000 and added a ticket for pagination.

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php (hunk 2 above)
Comment (bricht) [reply dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php]: Done

**Comment by bricht** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run Integration Tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

File: app/code/Magento/CatalogStorefront/Model/VariantService.php [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                $optionValues = $variantData->getOptionValues();
+                $id = $variantData->getId();
+                $explodedId = explode('/', $id);
+                $parentId = $explodedId[1];
+                $childId = $explodedId[2];
+                foreach ($optionValues as $optionValue) {
+                    preg_match('/(?<=:)(.*)(?=\/)/', $optionValue, $match);
+                    $attrCode = $match[0];
+                    $variant = [
+                        '_id' => $id . '/' . $attrCode,
```
Comment (mslabko) [inline app/code/Magento/CatalogStorefront/Model/VariantService.php:139]: why do we need custom value with attrCode for system "_id"?

**Comment by RuslanKostiv1** [comment]
@magento run Static tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

File: app/code/Magento/CatalogStorefront/Model/VariantService.php (hunk 5 above)
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/VariantService.php:139]: The variant is split up by option_values, and thus the variant_id is not unique anymore and would not work for _id

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run WebApi Tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run Static tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run WebAPI Tests

**Comment by bricht** [comment]
@magento run Integration Tests

**Comment by bricht** [comment]
@magento run Unit Tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by bricht** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Comment by RuslanKostiv1** [comment]
@magento import pr to magento-commerce/catalog-storefront-ce

**Comment by magento-engcom-team** [comment]
@RuslanKostiv1 the pull request successfully imported.

**Pull request merged**